go/oasis-node/cmd/storage/storage.go

```
@@ -395,15 +421,207 @@ func compactConsensusNodeDB(dataDir string) error {
+
+	if retainHeight <= startHeight {
+		logger.Info("consensus state already pruned", "retain_height", retainHeight, "start_height", startHeight)
+		return nil
+	}
+
+	logger.Info("pruning consensus state", "start_height", startHeight, "retain_height", retainHeight)
+	for h := startHeight; h < retainHeight; h++ {
+		if err := ndb.Prune(h); err != nil {
+			return fmt.Errorf("failed to prune version %d: %w", h, err)
+		}
+
+		if h%10_000 == 0 { // periodically sync to disk
+			if err := ndb.Sync(); err != nil {
+				return fmt.Errorf("failed to sync NodeDB: %w", err)
+			}
+			logger.Debug("forcing NodeDB disk sync during pruning", "version", h)
+		}
+	}
+
+	if err := ndb.Sync(); err != nil {
+		return fmt.Errorf("failed to sync NodeDB: %w", err)
+	}
+
+	return nil
+}
+
+// minReindexedHeight returns the smallest consensus height reindexed by any
+// of the configured runtimes.
+//
+// In case of no configured runtimes it returns max int64.
+func minReindexedHeight(dataDir string, runtimes []common.Namespace) (int64, error) {
+	fetchLastReindexedHeight := func(runtimeID common.Namespace) (int64, error) {
+		rtDir := runtimeConfig.GetRuntimeStateDir(dataDir, runtimeID)
+
+		history, err := history.New(runtimeID, rtDir, history.NewNonePrunerFactory(), true)
+		if err != nil {
+			return 0, fmt.Errorf("failed to open new light history: %w", err)
+		}
+		defer history.Close()
+
+		h, err := history.LastConsensusHeight()
+		if err != nil {
+			return 0, fmt.Errorf("failed to get last consensus height: %w", err)
+		}
+
+		return h, nil
+	}
+
+	var minH int64 = math.MaxInt64
+	for _, rt := range runtimes {
+		h, err := fetchLastReindexedHeight(rt)
+		if err != nil {
+			return 0, fmt.Errorf("failed to fetch last reindexed height for %s: %w", rt, err)
+		}
+
+		if h < minH {
+			minH = h
+		}
+	}
+
+	return minH, nil
+}
+
+func pruneCometDBs(dataDir string, retainHeight int64) error {
+	cmtConfig := cmtconfig.DefaultConfig()
+	_ = viper.Unmarshal(&cmtConfig)
```

Not sure why both `full` an `archive` node do viper unmarshal, nor why the error is ignored? So copying here but probably this should be fixed...

---

We can probably just remove this viper.Unmarshal here then?